### PR TITLE
Pulse 3618: Add limit messages to redis

### PR DESCRIPTION
**Pull request by rdempsey**
This PR adds the ability for Traptor to log limit message counts into Redis using an expiring key. The expiration time is the same as the rolling time window currently configured.

**Testing**

Get the code

- `git clone git@github.com:istresearch/traptor.git`
- `git checkout PULSE-3618-add-limit-messages-to-redis`

Set up your environment

- Create and activate a virtual environment (venv or Anaconda)
- Rename traptor.env.sample to traptor.env and fill in all fields. Some defaults are provided.
- Install the requirements: `pip install -r requirements.txt`

Run the tests

- In the root directory: `python -m pytest -vv --cache-clear --cov=traptor --cov-report html`

All tests should pass.

### CODE REVIEW

**Comment by madisonb** [comment]
@rdempsey can you commit/explain the reasoning behind using the limit messages as keys within redis? why not use a list, or store only the first limit message and the count/timestamp since?

**Comment by rdempsey** [comment]
@madisonb per the conversation with AY yesterday, in order to correctly create the multiple we need to apply to all of the rule stats to determine the "what should we have gotten rate" on a Traptor, we need the rate limit count at both the beginning and the end of the time window. It's the delta that will be used to create the multiplying factor that will be used. I need that information in Redis, and I need it on the same rolling time window as will be used for the stats gathering. Rather than simply the count of messages, I need the value (number of tweets held back) from the limit message itself. This does all of that.

**Comment by rdempsey** [comment]
@madisonb updated with your suggestions. Thanks for that! Much cleaner implementation.

File: traptor/traptor.py [hunk 1]
```diff
@@ -935,9 +955,10 @@ def run(self):
             self.twitter_rules = self._make_twitter_rules(self.redis_rules)
             self.logger.debug("Twitter rules: {}".format(self.twitter_rules.encode('utf-8')))
 
-            # Make the rule counters
+            # Make the rule and limit message counters
             if self.traptor_type != 'locations':
                 self._make_rule_counters()
+                self._make_limit_message_counter()
```
Comment (andrewkcarter) [inline traptor/traptor.py]: Does Twitter API guarantee we'll never see a limit type message for location traptors? If not, then we have possible `NoneType` conditions and we should either alway init the counters, or do existence checks before we try to use them.

Comment (rdempsey) [inline traptor/traptor.py]: We receive a lot of limit messages from location Traptors. However, this information is being used for rule rebalancing, which only applies to follow and track traptors. I will add the existence check for the counter to ensure we don't run into the `NoneType` issue you mention. Thanks!

**Pull request merged**